haystack/pipeline_eval.py

```diff
@@ -0,0 +1,265 @@
+    )
+    if status.returncode:
+        logger.warning("Tried to start Elasticsearch through Docker but this failed. "
+                       "It is likely that there is already an existing Elasticsearch instance running. ")
+    else:
+        time.sleep(15)
+
+def main():
+
+    launch_es()
+
+    document_store = ElasticsearchDocumentStore()
+    es_retriever = ElasticsearchRetriever(document_store=document_store)
+    eval_retriever = EvalRetriever()
+    reader = FARMReader("deepset/roberta-base-squad2", top_k_per_candidate=4, num_processes=1, return_no_answer=True)
+    eval_reader = EvalReader(debug=True)
+
+    # Download evaluation data, which is a subset of Natural Questions development set containing 50 documents
+    doc_dir = "../data/nq"
+    s3_url = "https://s3.eu-central-1.amazonaws.com/deepset.ai-farm-qa/datasets/nq_dev_subset_v2.json.zip"
+    fetch_archive_from_http(url=s3_url, output_dir=doc_dir)
+
+    # Add evaluation data to Elasticsearch document store
+    # We first delete the custom tutorial indices to not have duplicate elements
+    preprocessor = PreProcessor(split_length=500, split_overlap=0, split_respect_sentence_boundary=False, clean_empty_lines=False, clean_whitespace=False)
+    document_store.delete_all_documents(index=doc_index)
+    document_store.delete_all_documents(index=label_index)
+    document_store.add_eval_data(filename="../data/nq/nq_dev_subset_v2.json", doc_index=doc_index, label_index=label_index, preprocessor=preprocessor)
+    labels = document_store.get_all_labels(index=label_index)
+    q_to_l_dict = {x.question: x.answer for x in labels}
+
+    # Here is the pipeline definition
+    p = Pipeline()
+    p.add_node(component=es_retriever, name="ESRetriever", inputs=["Query"])
+    p.add_node(component=eval_retriever, name="EvalRetriever", inputs=["ESRetriever"])
+    p.add_node(component=reader, name="QAReader", inputs=["EvalRetriever"])
+    p.add_node(component=eval_reader, name="EvalReader", inputs=["QAReader"])
+
+    results = []
+    for i, (q, l) in enumerate(q_to_l_dict.items()):
+        res = p.run(query=q,
+                    top_k_retriever=top_k_retriever,
+                    labels=l,
+                    top_k_reader=10,
+                    index=doc_index,
+                    # skip_incorrect_retrieval=True
+                    )
+        results.append(res)
+
+    eval_retriever.print()
+    print()
+    es_retriever.print_time()
+    print()
+    eval_reader.print(mode="reader")
+    print()
+    reader.print_time()
+    print()
+    eval_reader.print(mode="pipeline")
+
+
+class EvalRetriever:
+    def __init__(self, debug=False):
+        self.outgoing_edges = 1
+        self.correct_retrieval_count = 0
+        self.query_count = 0
+        self.has_answer_count = 0
+        self.has_answer_correct = 0
+        self.has_answer_recall = 0
+        self.no_answer_count = 0
+        self.recall = 0.0
```

Comment:
I agree that mAP would be useful. I think this would be best tackled when we look at how we integrate more metrics (see #925) since this will define the structure of the labels for the retriever.

If you'd like to measure mAP already, we can certainly discuss how to implement a short term hack

Reply:
Sure, makes sense to wait until a general solution is worked on. I'll ping you on Slack regarding a short term hack to avoid cluttering your PR 😃